qiskit/qpy/__init__.py

```diff
@@ -382,31 +382,58 @@ def open(*args):
+classes.
+
+DURATION
+~~~~~~~~
+
+A :class:`~.circuit.Duration` is encoded by a single-byte ASCII ``char`` that encodes the kind of
+type, followed by a payload that varies depending on the type.  The defined codes are:
+
+==============================  =========  =========================================================
+Qiskit class                    Type code  Payload
+==============================  =========  =========================================================
+:class:`~.circuit.Duration.dt`   ``t``     One ``unsigned long long value``.
+
+:class:`~.circuit.Duration.ns`   ``n``     One ``double value``.
+
+:class:`~.circuit.Duration.us`   ``u``     One ``double value``.
+
+:class:`~.circuit.Duration.ms`   ``m``     One ``double value``.
+
+:class:`~.circuit.Duration.s`    ``s``     One ``double value``.
+
+==============================  =========  =========================================================
 
 Changes to EXPR_TYPE
 ~~~~~~~~~~~~~~~~~~~~
 
 The following table shows the new type classes added in the version:
 
-======================  =========  =================================================================
-Qiskit class            Type code  Payload
-======================  =========  =================================================================
-:class:`~.types.Float`  ``f``      None.
-======================  =========  =================================================================
+=========================  =========  ==============================================================
+Qiskit class               Type code  Payload
+=========================  =========  ==============================================================
+:class:`~.types.Float`     ``f``      None.
+
+:class:`~.types.Duration`  ``d``      None.
+
+=========================  =========  ==============================================================
 
 Changes to EXPR_VALUE
 ~~~~~~~~~~~~~~~~~~~~~
 
 The classical expression's type system now supports new encoding types for value literals, in
 addition to the existing encodings for int and bool. The new value type encodings are below:
 
-===========  =========  ============================================================================
-Python type  Type code  Payload
-===========  =========  ============================================================================
-``float``    ``f``      One ``double value``.
+===========================  =========  ============================================================
+Python type / class          Type code  Payload
+===========================  =========  ============================================================
+``float``                    ``f``      One ``double value``.
 
-===========  =========  ============================================================================
+:class:`~.circuit.Duration`  ``t``      One ``DURATION``.
+
+===========================  =========  ============================================================
```

Comment:
`circuit.Duration` is still a Python type, tbf

I think there's a risk with using `t`: I think that type code is already used elsewhere to represent "tuple", and there's a risk of a clash (despite the organisational structure on `type_keys.py`, the enums in there are _not_ necessarily separate namespaces in QPY land).  That said, I think `EXPR_VALUE` is localised enough that it can never hit the clash, but `PARAM_TYPE` would.  Up to you if you think it's worth changing.

Reply:
I might not be understanding. Are you saying that we should try to have unique characters across all of `type_keys.py` in case we introduce a bug that reads/writes a byte in the wrong place? My understanding of these is that we're manually reading / writing a byte that can only be one of e.g. `TypeKeys.ExprType.<KEYS>` in the place that we're reading / writing.

Reply:
I can't remember the exact circumstances, but I remember having trouble with QPY at one point because there were locations where type keys from more than one `TypeKeys.<Thing>` were valid choices - I think the abstraction is leaky.

That said, given that I added the `ExprType` ones after that, I'm relatively sure that it's not a problem here, so it's ok to leave.